src/main/scala/decider/Z3ProverAPI.scala

```diff
@@ -0,0 +1,439 @@
+  val emittedFuncs = mutable.LinkedHashSet[FuncDecl]()
+  val emittedFuncSymbols = mutable.Queue[Symbol]()
+
+  val expandMacros = true
+
+
+
+  def version(): Version = {
+    var versString = com.microsoft.z3.Version.getFullVersion
+    if (versString.startsWith("Z3 "))
+      versString = versString.substring(3)
+    Version(versString)
+  }
+
+  def start(): Unit = {
+    pushPopScopeDepth = 0
+    lastTimeout = -1
+    logfileWriter = if (Verifier.config.disableTempDirectory()) null else viper.silver.utility.Common.PrintWriter(Verifier.config.proverLogFile(uniqueId).toFile)
+    ctx = new Context(Z3ProverAPI.initialOptions.asJava)
+    val params = ctx.mkParams()
+    Z3ProverAPI.boolParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    Z3ProverAPI.intParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    Z3ProverAPI.doubleParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    Z3ProverAPI.stringParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    prover = ctx.mkSolver()
+    prover.setParameters(params)
+    termConverter.start()
+    termConverter.ctx = ctx
+    emittedPreambleString.clear()
+    preamblePhaseOver = false
+    emittedFuncs.clear()
+    emittedSorts.clear()
+    emittedFuncSymbols.clear()
+    emittedSortSymbols.clear()
+  }
+
+  def reset(): Unit = {
+    stop()
+    termConverter.reset()
+    start()
+  }
+
+  def stop(): Unit = {
+    if (ctx != null){
+      ctx.close()
+      ctx = null
+      emittedPreambleString.clear()
+      preamblePhaseOver = false
+      emittedFuncs.clear()
+      emittedSorts.clear()
+      emittedFuncSymbols.clear()
+      emittedSortSymbols.clear()
+    }
+  }
+
+  def push(n: Int = 1): Unit = {
+    endPreamblePhase()
+    pushPopScopeDepth += n
+    if (n != 1)
+      sys.error("Not supported")
+    prover.push()
+  }
+
+  def pop(n: Int = 1): Unit = {
+    endPreamblePhase()
+    prover.pop(n)
+    pushPopScopeDepth -= n
+  }
+
+  def emit(content: String): Unit = {
+    if (preamblePhaseOver) {
+      sys.error("emitting phase over")
+    }
+    emittedPreambleString.append(content)
+  }
+
+  override def emit(contents: Iterable[String]): Unit = {
+    if (preamblePhaseOver) {
+      sys.error("emitting phase over")
+    }
+    emittedPreambleString.appendAll(contents)
+  }
+
+  override def emitSettings(contents: Iterable[String]): Unit = {
+    // we ignore this, don't know any better solution atm.
+    // our settings are defined in this class (see above).
+    // except we check if someone gives us our custom randomization string, in which
+    // case we'll directly set the options to randomize.
+    val optionSetString = s"(set-option :${Z3ProverAPI.randomizeSeedsSetting}"
+    if (contents.exists(s => s.startsWith(optionSetString))) {
+      val params = ctx.mkParams()
+      Z3ProverAPI.randomizeSeedsOptions.foreach(o => params.add(o, Random.nextInt(10000)))
+      prover.setParameters(params)
+    }
```

Comment:
Can we emit warnings for ignored settings?

Reply:
Z3 will output an error and potentially an exception when you give it settings that it doesn't recognize via API.